### PR TITLE
add-privileged-users-in-room

### PR DESCRIPTION
**Pull request by GoodGuyMarco**
Jira-Ticket: https://element-io.atlassian.net/browse/PSB-169
Related PR: https://github.com/matrix-org/matrix-js-sdk/pull/2892.
---
Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

OSS_1

When searching users via the new `AutocompleteInput` component, suggestions are shown and can be selected. Select a power level and hit the "Apply" button to confirm your changes. Users should appear in the "Privileged Users" section afterwards.

OSS_2

## Checklist

* [x] Tests written for new code (and old code if feasible)
* [x] Linter and other CI checks pass
* [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

Notes: Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

element-web notes: Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions.

Signed-off-by: Marco Bartelt marco.bartelt.ext@bwi.de

---
Here's what your changelog entry will look like:

## ✨ Features
 * Adds a new section under "Room Settings" > "Roles & Permissions" which adds the possibility to multiselect users from this room and grant them more permissions. ([\#9596](https://github.com/matrix-org/matrix-react-sdk/pull/9596)). Contributed by @GoodGuyMarco.

### CODE REVIEW

**Comment by t3chguy** [comment]
@jakewb-b could you get a designed tasked with reviewing it?

**Comment by jakewb-b** [comment]
> @jakewb-b could you get a designer tasked with reviewing it?

Yup - already been flagged to @gaelledel who was also involved in the original discussions on this.

**Review by gsouquet — CHANGES_REQUESTED** [review]
Thank you for opening this pull request!
I've added a few comments that would need to be adressed before this can be merged in. There's also a few issue in regards to TypeScript strict mode that have been picked up by our CI that would need your attention.

I have not yet checked the tests you added, but will do once the rest of the code reaches it's more final shape

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: Fixed in this commit https://github.com/matrix-org/matrix-react-sdk/pull/9596/commits/01220e9a653b216944d01f43fd9eb539944fae96

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: I have replaced that by `$secondary-content` 🧑‍🎨

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: I tried replacing all hardcoded instances with the spacing variables. Point me out if I missed any 🕵️

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: Can you clarify what you mean by that?

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: I agree. Could you point me to a combination of background and font color variables that I should use for this? 

pill

File: src/i18n/strings/de_DE.json
Comment (GoodGuyMarco) [reply src/i18n/strings/de_DE.json]: Didn't know that. Removed the german translations ✅

File: src/components/views/settings/AddPrivilegedUsers.tsx
Comment (GoodGuyMarco) [reply src/components/views/settings/AddPrivilegedUsers.tsx:83]: There was no `form` element until now. I've introduced one, but that leads to not beeing able to use the `SettingsFieldSet` component as it does not work well when beeing nested inside a `form`. I've reused the classes of the component but am unsure if this is a good solution. Any ideas?

File: src/components/views/settings/AddPrivilegedUsers.tsx
Comment (GoodGuyMarco) [reply src/components/views/settings/AddPrivilegedUsers.tsx]: Done in https://github.com/matrix-org/matrix-react-sdk/pull/9596/commits/e3d2113d5d094ff396d4f3d44ffb4dbd7b25779a

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx:51]: Tried that but apparrently the suggestions don't close when clicking away from the component 🤔

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx]: Yeah that's odd indeed (code was not written by me initially). I've changed the component to be controlled for `selection` and `onSelectionChange` as this makes it far more easy in my opinion.

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx]: I agree, but now it is kind of obsolet because of https://github.com/matrix-org/matrix-react-sdk/pull/9596#discussion_r1028956386

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx]: I've changed the behaviour to work with a boolean `isFocused`.

File: src/components/views/settings/AddPrivilegedUsers.tsx
Comment (GoodGuyMarco) [reply src/components/views/settings/AddPrivilegedUsers.tsx]: EDIT: Found an alternative solution using flexbox which is able to reuse the `SettingsFieldSet` component.

**Comment by GoodGuyMarco** [comment]
Hey @gsouquet, thank you for your constructive feedback! I tried to address all of your comments. Some of them were fixable, others need your feedback.  

Could you help me regarding the TypeScript strict mode issue? Yesterday it showed me only errors in my changed files. Today it gives me 600 > errors, which is confusing. Maybe I could solve it with a rebase, but I'm not sure. Anyway it is not really convenient that I am not able to run this check locally like the Linters because that heavily increases the time spend on the feedback loop. Any ideas?

File: res/css/structures/_AutocompleteInput.pcss
Comment (gsouquet) [reply res/css/structures/_AutocompleteInput.pcss]: Having `!important` is usually an anti-pattern and this code is using a lot of it... which is slightly worrying and might be a sign that we're abusing another component in fitting these requirements

File: res/css/structures/_AutocompleteInput.pcss
Comment (gsouquet) [reply res/css/structures/_AutocompleteInput.pcss]: The @matrix-org/design review should point you in the right direction

File: src/components/structures/AutocompleteInput.tsx
Comment (gsouquet) [inline src/components/structures/AutocompleteInput.tsx:200]: We actually have a react hook called `useFocus` that you could use here.

**Review by gsouquet — CHANGES_REQUESTED** [review]
Thanks for adressing the feedback, a few minor things here. But this is definitely on the right track!
Next step would be to get a review from design.

A rebase with `develop` should help with CI going a bit crazy with the TS issues. If you want to test this locally, you can enable `strict: true` in `tsconfig.json` temporarily

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx:200]: Neat, thanks for the hint 👍

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss]: I've investigated the styling and found that global input styles are interfering with these styles. I've tried to get rid of some `!important`'s but did not manage to do that without breaking component encapsulation. The input styles in this component are kind of special as the input needs to be hidden. 

Maybe we could use a different approach where we make the actual input visible for the user, position the selection pills absolute above the input and dynamically increase the padding on the left, but I don't know if that is worth the effort to get rid of some `!important`'s.

File: src/components/structures/AutocompleteInput.tsx
Comment (GoodGuyMarco) [reply src/components/structures/AutocompleteInput.tsx:114]: I agree and changed it as suggested!

File: res/css/structures/_AutocompleteInput.pcss
Comment (GoodGuyMarco) [reply res/css/structures/_AutocompleteInput.pcss:1]: Upsi 👀

**Comment by GoodGuyMarco** [comment]
@gsouquet Hey, I've just rebased against `develop` and addressed your second feedback loop last week. Could you have another look and ask for the design feedback?

**Review by gaelledel — APPROVED** [review]
looks good to me thanks!

**Review by gsouquet — APPROVED** [review]
Code looks good to me, there are still some design outstanding question for @gaelledel before we can merge this.

**Comment by GoodGuyMarco** [comment]
Thank you for your feedback @gaelledel. I have a few questions, as speaking about visuals always leaves room for misunderstanding.

> 1. For the search users field: We need to add a proper search field component with the Search icon and the following string "Search users"

1. By _"...search field component with the search icon..."_ you mean that the input should have a magnifier icon on the left side (like in the screenshot)?

search

2. By _"...and the following string "Search users"..."_ you want to replace the placeholder I suggested `Search users in this room...` by `Search users`, correct?

---

> 2. Search users field: As soon as the user clicks in the field, we should display results in the drop down. Currently it doesn't show anything

Currently it displays results (matching your search criteria) as soon as you start typing. Why do we think that it should appear right away and what are your expectations if we change that regarding following questions: 

1. How many results should appear when the room has many users. Should there be a maximum and what should it be?
2. Which users should appear first when there is no search criteria?

---

> 3. Search field select colour should be set to Quarternary

Do you mean the background color of a selected suggestion?

suggestion

---

> 4. When selecting "custom level" and clicking away from the fields it doesn't deselect the option which as a user does not allow me to rectify my choice to change a permission. No matter what I do, the Custom level is set to changeable numbers and does not give me the ability to change to another permission type

Do you mean the the user is "trapped" inside the "custom level" input and can't go back to the select field? This is actually the current behaviour of this component. I stumbled across this aswell, but would suggest not changing this in my PR as it seems unrelated and deserves another issue.

**Comment by GoodGuyMarco** [comment]
@gaelledel 

### 🧑‍🎨 Added a magnifier on the left
---
OSS_i1

### 🧑‍🎨 Aligned look and feel to regular input components (focus border color, equal height)
---
OSS_i2

### 🧑‍🎨 Updated hover and selected background color of the suggestions
---
OSS_i3

**Review by gaelledel — APPROVED** [review]
Reviewed with Marco and all good from me

**Review by gsouquet — APPROVED** [review]
Looks good to me, but the CI is not too happy. Let's fix that, and we should be able to merge this change in!

**Comment by t3chguy** [comment]
Branch matching is likely kicking in, their js-sdk needs updating or deleting

**Comment by GoodGuyMarco** [comment]
@t3chguy I just rebased the PR. What do you mean by "branch matching"?

**Comment by t3chguy** [comment]
Our CI attempts to pick matching branches between all the various layers (js-sdk, react-sdk, element-web, etc) by finding branches of the same name in your forks. https://github.com/matrix-org/matrix-js-sdk/pull/2892 is such a branch, and that is based on an older version of develop so the branch needs either updating or deleting.

**Comment by GoodGuyMarco** [comment]
@t3chguy Okay. I could not find the merged branch in matrix-js-sdk. Did you delete it already? If so, could you manually re-run the CI to see if that helped?

**Comment by t3chguy** [comment]
We can't delete branches on your fork

**Comment by GoodGuyMarco** [comment]
@t3chguy my bad, deleted the branch in our fork ✅

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:35]:
```suggestion
    beforeEach(() => {
```

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:40]: `getBy` already ensures that this element exists and explodes otherwise.
If you want to test, that some element exist you could do

```
expect(screen.queryByLabelText("something")).toBeInTheDocument();
```

File: src/components/structures/AutocompleteInput.tsx
Comment (weeman1337) [inline src/components/structures/AutocompleteInput.tsx:147]: We should better test by using labels / texts.

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:63]: We should use `type` here. Like mentioned in other places we should stay close on how users are acting. Typing into the field would be an option https://testing-library.com/docs/ecosystem-user-event/#typeelement-text-options

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:66]: We should check for the expected texts to be visible here.

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:84]: Same as in other places, we should better check for texts, labels etc..

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:46]:
```suggestion
    it('should render suggestions when a query is set', () => {
```

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:70]:
```suggestion
    it('should render selected items passed in via props', () => {
```

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:88]:
```suggestion
    it('should call onSelectionChange() when an item is removed from selection', () => {
```

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [inline test/components/structures/AutocompleteInput-test.tsx:113]:
```suggestion
    it('should render custom selection element when renderSelection() is defined', () => {
```

File: test/components/views/settings/AddPrivilegedUsers-test.tsx
Comment (weeman1337) [inline test/components/views/settings/AddPrivilegedUsers-test.tsx:91]: Can this be merged to the previous `act`?

**Review by weeman1337 — COMMENTED** [review]
Thanks for adding the tests. Tests should be more focused to mimic real user interaction like looking for texts / labels and typing / clicking instead of triggering change events.

File: test/components/structures/AutocompleteInput-test.tsx
Comment (weeman1337) [reply test/components/structures/AutocompleteInput-test.tsx:66]: Can be improved later

File: test/components/views/settings/AddPrivilegedUsers-test.tsx
Comment (GoodGuyMarco) [reply test/components/views/settings/AddPrivilegedUsers-test.tsx:99]: Actually started with a merged `act` and divided it because it failed 🤔

Comment (weeman1337) [reply test/components/views/settings/AddPrivilegedUsers-test.tsx:99]: Ok, lets keep it for now

**Review by weeman1337 — APPROVED** [review]

**Pull request merged**
